src/resources/spi_flash_esp32.cc

```diff
@@ -0,0 +1,285 @@
+  if (ret != ESP_OK) {
+    group->close();
+    return Primitive::os_error(ret, process);
+  }
+
+  group->set_data_partition(partition);
+
+  esp_vfs_fat_mount_config_t mount_config = {
+      .format_if_mount_failed = static_cast<bool>(format_if_mount_failed),
+      .max_files = max_files,
+      .allocation_unit_size = static_cast<size_t>(allocation_unit_size)
+  };
+
+  wl_handle_t wl_handle;
+  ret = esp_vfs_fat_spiflash_mount(mp, mount_point, &mount_config, &wl_handle);
+  if (ret != ESP_OK) {
+    group->close();
+    return Primitive::os_error(ret, process);
+  }
+
+  group->set_wl_handle(wl_handle);
+
+  proxy->set_external_address(group);
+  return proxy;
+}
+
+PRIMITIVE(init_nand_flash) {
+#ifdef CONFIG_SPI_FLASH_NAND_ENABLED
+  ARGS(cstring, mount_point, SPIResourceGroup, spi_bus, int, gpio_cs, int, frequency, int, format_if_mount_failed, int, max_files, int, allocation_unit_size);
+
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (!proxy) ALLOCATION_FAILED;
+
+  char* mp = static_cast<char *>(malloc(strlen(mount_point)));
+  if (!mp) MALLOC_FAILED;
+  strcpy(mp, mount_point);
+
+  auto group = _new SPIFlashResourceGroup(process,  mp);
+
+  if (!group) {
+    free(mp);
+    ALLOCATION_FAILED;
+  }
+
+  spi_device_interface_config_t dev_cfg = {
+      .mode = 0,
+      .clock_speed_hz = frequency,
+      .spics_io_num = gpio_cs,
+      .flags = SPI_DEVICE_HALFDUPLEX,
+      .queue_size = 1
+  };
+  ESP_LOGI("toitspi", "Setting up device with frequency: %d", frequency);
+  spi_device_handle_t nand_spi_device;
+  esp_err_t ret = spi_bus_add_device(SPI3_HOST, &dev_cfg, &nand_spi_device);
+  if (ret != ESP_OK) {
+    group->close();
+    return Primitive::os_error(ret, process);
+  }
+  group->set_nand_spi_device(nand_spi_device);
+
+  spi_nand_flash_config_t nand_config = {
+      .device_handle = nand_spi_device,
+      .gc_factor = 45
+  };
+  ESP_LOGI("toitspi", "Init nand flash device");
+  spi_nand_flash_device_t *nand_flash_device;
+  ret = spi_nand_flash_init_device(&nand_config, &nand_flash_device);
+  if (ret != ESP_OK) {
+    group->close();
+    return Primitive::os_error(ret, process);
+  }
+  group->set_nand_flash_device(nand_flash_device);
+
+  ESP_LOGI("toitspi", "Mount allocation unit size: %d, max_files: %d",allocation_unit_size, max_files);
+  esp_vfs_fat_mount_config_t mount_config = {
+      .format_if_mount_failed = static_cast<bool>(format_if_mount_failed),
+      .max_files = max_files,
+      .allocation_unit_size = static_cast<size_t>(allocation_unit_size)
+  };
+
+  ret = esp_vfs_fat_nand_mount(mp, nand_flash_device, &mount_config);
+  if (ret != ESP_OK) {
+    group->close();
+    return Primitive::os_error(ret, process);
+  }
+
+  proxy->set_external_address(group);
+  return proxy;
+#else
+  UNIMPLEMENTED_PRIMITIVE;
+#endif // CONFIG_SPI_FLASH_NAND_ENABLED
+
+}
+
+PRIMITIVE(close) {
+  ARGS(SPIFlashResourceGroup, group)
+  group->close();
```

Comment:
Remember to clear the external address of the proxy.

```suggestion
  group->close();
  group_proxy->clear_external_address();
```

Reply:
I am still not clear why that is necessary. The proxy would be "dead" anyway in the toit vm and would be gc'ed soon after. What does the clear_external_address accomplish? Isn't it equivalent to setting some of the private fields to null in the destructor that you commented on?

Reply:
The proxy object is still in the object heap, so it is still reachable after you've returned from the primitive. If you're being a good sport, you'll probably get rid of the references to the proxy object at some point, but in the mean time it isn't great to have a dangling pointer to freed memory.

Reply:
So, this is to catch "access after free" of resources (or is it to ensure idempotency of close? Then a check should be made for the value of group not being null). That is almost similar to setting private pointer members of classes to null in destructors. There is only a subtle difference in that the proxy is still kind-of-live, whereas a freed object is not.